### DISCUSSION TITLE
PyG heterogeneous graph includes the global node id?

### DISCUSSION THREAD

**Discussion by jiaqian** [pyg-team/pytorch_geometric#5304] [Q&A]
Hi all,

I want to add the **global node id** in each batch (on heterogeneous graph) for some post analysis. I know `Data` object can easily do it by adding `data.n_id = torch.arange(data.num_nodes)`. I tried to add a dictionary by specifying different types of nodes, e.g., `Hetero_data.n_id = {"type1":torch.arange(Hetero_data.x_dict["type1"].shape[0]),"type2":...}`. In each batch, it has the all the original node ids instead of the ones only in this batch.

Anyone has idea how to do it? Any suggestion is welcome, thanks in advance.

**Comment by rusty1s** [comment]
For heterogeneous graphs, you would do this per node type:
```python
for node_type in data.node_types:
    data[node_type].n_id = torch.arange(data[node_type].num_nodes)

**Reply by jiaqian** [reply]
it works, thanks.

**Reply by jiaqian** [reply]
Hi, in my case I work on link prediction task and the label (another type of edge) is also included in `edge_index_dict` in `HeteroData`. 

1. Does it mean during aggregation process, the label edge information has been used as well? 
2. If this is the case, is there any way not using it during aggregation, but I need to use label edge for `decoder`(take the element-wise product and supply to a fully-connected layer) to train the model in a supervised way.

Thanks in advance.

**Reply by rusty1s** [reply]
Likely. It is best to ensure that `edge_index` and `edge_label_index` do not share information. You can implement this as a simple pre-processing before model training.

**Reply by jiaqian** [reply]
Thanks for the reply. Is it also possible to use the `HeteroConv` wrapper and only specify the (graph structure) edge type there rather than label edge? I think that might be easier since I can still use dataloader to batch the label edges for training.

**Reply by rusty1s** [reply]
You mean to exclude the edge label type as part of the edge types? This should be possible by simply removing this information from the `metadata`.

**Reply by jiaqian** [reply]
ok, I see. Thanks.
